Data Access Layer

colewinfield opened the pull request:
Moved ICardService and Models to a Data Access Layer (out of the Presentation Layer). UI now interacts with ViewModels. 

@rlayman87 and @dyohn commented on having data access in the UI layer. This is my first time refactoring architecture like this. Would this follow the MVC pattern more appropriately? If not, please leave your feedback.

Freytes committed:
Upgraded .NET level to 6.0 for database used MongoDB drivers.

Added the following DB Object fields:

- Subject
- Question
- Answer
- Is Custom
- Difficulty Level
- Notes

colewinfield committed:
Begin refactor of database

This commit is the start of refactoring the database implementation; currently, it isn't abstracted and using DI.

colewinfield committed:
Refactor CardController for succinctness

colewinfield committed:
Finish DI refactor for DB

Freytes committed:
Merge pull request #12 from dyohn/CRUDFeatures_refactoring

Crud features refactoring

Freytes committed:
Updated the database to include more fields for user input, display and edit.

colewinfield committed:
Add carousel to HomePage

colewinfield committed:
Fix merge conflict in Startup.cs

colewinfield committed:
Add Model data to Index

colewinfield committed:
Finish carousel

colewinfield committed:
Fix carousel card-title font size

Changed it to be smaller: 2.5 -> 1.5rem

colewinfield committed:
Merge branch 'layout_homepage_carousel' into layout_homepage

colewinfield committed:
Start DAL refactor

Create a third layer for data access management. Removing database access from the UI layer. Replacing models within UI with ViewModels.

colewinfield committed:
Finish refactror of Data Layer

Moved to a ViewModel pattern for UI models. Created a Data Access Layer away from the UI. Controller now pass an instance of VMs to Views.

colewinfield committed:
Use ViewModel in CardController

Forgot to switch from Enumerable<Card> to CardViewModel in CardController. This fixes the bug.

colewinfield committed:
Move Repository project back to src

Somehow, it got moved to outside of /src/ and was in the parent directory. Moved it back to where it belongs and readded the dependencies.

jonnywwright commented on src/Mentat.Repository/Models/Card.cs:
```diff
@@ -8,6 +8,16 @@ namespace Mentat.UI.Models
+
+        public string difficulty_level { get; set; }
+
         public string notes { get; set; }
```
Properties should be pascal-case

jonnywwright commented on src/Mentat.UI/Views/Home/Index.cshtml:
```diff
@@ -1,12 +1,59 @@
+              <!-- Card Indicator -->
+              <ul class="carousel-indicators">
+                @{int j = 0;}
+                @foreach (var item in Model.cards)
```
Can you do a regular for-loop? Applies to both for-eaches

colewinfield replied:
@jonnywwright, Out of curiosity, is there something wrong with a for-each-loop?

jonnywwright replied:
Declaring an index outside a foreach is typically a hint you should be using a classic for loop. You've elected to use a List to hold your cards, therefore you have access to Count and index accessing. A for-loop is a better fit.

jonnywwright commented on src/Mentat.UI/ViewModels/CardViewModel.cs:
```diff
@@ -0,0 +1,38 @@
+        public string notes { get; set; }
+
+        public List<Card> cards { get; set; }
+        public CardViewModel(Card card)
```
Properties should be pascal case

jonnywwright commented on src/Mentat.UI/ViewModels/CardViewModel.cs:
```diff
@@ -0,0 +1,38 @@
+        }
+        public CardViewModel(List<Card> cards)
+        {
+            this.cards = cards;
```
after pascal casing the properties remove 'this.'

jonnywwright commented on src/Mentat.UI/Controllers/HomeController.cs:
```diff
@@ -6,21 +6,52 @@
+                    indices.Add(index);
+                    randomCards.Add(cards[index]);
+                }
+            }
```
Import the namespace System.Linq;

Replace line 32 thru 46 with

```csharp
return cards.OrderBy(a => Guid.NewGuid()).Take(numCards);
```

jonnywwright replied:
side note, don't use the letter 'a' in my example (I copy pasted that). Use 'card', and if you don't switch the return to IEnumerable<Card> you will need to also add a ToList() after the Take

jonnywwright commented on src/Mentat.UI/Controllers/CardController.cs:
```diff
@@ -48,27 +48,27 @@ public ActionResult Create()
-                               Models.MongoHelper.database.GetCollection<Models.Card>("FlashCards");
-                
+            {                
                 object id = GenRandomId(24);
```
Not sure what this id is for but would Guid.NewGuid() suffice?

jonnywwright commented on src/Mentat.Repository/Models/Card.cs:
```diff
@@ -8,6 +8,16 @@ namespace Mentat.UI.Models
+        public string difficulty_level { get; set; }
+
         public string notes { get; set; }
     }
```
Needs pascal casing. Typically you shouldn't snake case in C# unless its a serialization requirement to be compatible with something else.

jonnywwright commented on src/Mentat.Repository/Models/CardDatabaseSettings.cs:
```diff
@@ -0,0 +1,9 @@
+        public string ConnectionString { get; set; } = "mongodb+srv://cdb:eN9qAv7forAKSnEt@flashcardapp.4yggo3z.mongodb.net/?retryWrites=true&w=majority";
+        public string DatabaseName { get; set; } = "fcApp";
+        public string CardCollectionName { get; set; } = "FlashCards";
+    }
```
Can you put a white space between properties like you do with other classes?

jonnywwright commented on src/Mentat.Repository/Models/ICardDatabaseSettings.cs:
```diff
@@ -0,0 +1,9 @@
+        string CardCollectionName { get; set;  }
+        string DatabaseName { get; set; }
+        string ConnectionString { get; set; }
+    }
```
Can you put a white space between properties like you do with other classes?

jonnywwright commented on src/Mentat.Repository/Services/ICardService.cs:
```diff
@@ -0,0 +1,13 @@
+        Card Create(Card card);
+        void Update(string id, Card card);
+        void Remove(string id);
+    }
```
Add some whitespace. Consider changing all Lists to IEnumerable.

jonnywwright commented on src/Mentat.UI/Views/Home/Index.cshtml:
```diff
@@ -1,12 +1,59 @@
+                  @{int i = 0;}
+                  @foreach (var item in Model.cards)
+                 {
+                    i++;
```
This should go at the end of the loop like the other one. Be consistent. Better yet do a for-loop

jonnywwright commented on src/Mentat.Repository/Models/Card.cs:
```diff
@@ -8,6 +8,16 @@ namespace Mentat.UI.Models
     public class Card
     {
         public Object _id { get; set; }
+        public string subject { get; set; }
```
need a white space after _id

rlayman87 commented:
Hey, @colewinfield
I noticed that all of your files you moved to the Repository still have a namespace of Mentat.UI instead of Mentat.Repository....
That kind of thing is easily missed, though. I've done it myself in the past. Other than that, I do believe that was what @dyohn  was wanting.
Did you have any thoughts on making the method names for the Service more descriptive? If not, that's fine. Just thought I would ask.
One other thing on your carousel code is, visually, I would suggest making the carousel wider to accommodate longer questions and answers. If you have time. If not, I think putting it as a bug in Trello would suffice.
![image](https://user-images.githubusercontent.com/55672598/178646704-b0d5bf4f-48b6-44e1-9e27-65fd375ffe87.png)

Rachel

colewinfield committed:
Change namespaces.

colewinfield commented:
@rlayman87, thank you for reviewing. Totally forgot to change the namespaces. But I've updated those. Also, I've changed the Service's CRUD method names to better descriptors. As for the carousel, I'll make the card wider - but I'll also likely add a limit to the number of characters viewable in the next push. Thanks again for looking over the refactor.

colewinfield committed:
Fix round of PR review updates

Fix some PascalCasing, spacing, and other conventions. Also reduced a randomizer to Linq line.

colewinfield committed:
Update foreach loop to for-loop in Home Index

colewinfield committed:
Changing PascalCase back to database naming scheme

When retrieving the data from Mongo, I believe the Model's properties have to correspond with what's retrieved from the DB. Chris will have to rename these variables (I don't have access).

colewinfield committed:
Add limit to word-count for carousel-card

Still needs some work (perhaps limit the characters). I'll leave it for the next chump.

colewinfield committed:
Merge conflict resolve

Fixed conflict in appsettings.json

colewinfield commented:
@jonnywwright, I fixed the spacing issues you've mentioned and refactored the for-loops in the Razor HTML. I've also changed the method of randomly picking the cards with the LINQ query (that was very clever). And the cards that are Created hereafter use the Guid.NewGuid() method. I initially changed the properties to PascalCase, but it created conflicts with the databasing. I didn't set up the MongoDB, so I'm assuming the properties have to match the returned names of the elements from the database. I ran into issues when they weren't named similarly (I'm assuming they were named this way by @Freytes to correspond with the DB). 

@rlayman87 I've changed the card-width and added a JS script to limit the number of words displayed per card. It probably should be the number of characters, but I'm beat. 

The merge conflict in the appsettings.json is also resolved.

colewinfield committed:
Merge branch 'development' into data_layer

Commit:
Changed CardViewMode propertiesl to pascal-case.

Commit:
Added property serialization on the card dto to allow for pascal-casing.

jonnywwright commented:
@colewinfield I pulled your branch and fixed the casing issues. Please review the changes and merge them into this branch.
https://github.com/dyohn/mentat/pull/25

Commit:
More corrections on the card class.

Commit:
Fixed ordering of attributes on the Card Id property.

Commit:
Removed whitespace and formatting

Commit:
Formatting

colewinfield committed:
Merged fixes and refactor site.js

Merged in JW's PascalCase fixes and created a function in site.js.

colewinfield committed:
Merge pull request #25 from dyohn/feature/data_layer/fix-casing

Feature/data layer/fix casing

colewinfield committed:
Merge branch 'data_layer' of https://github.com/dyohn/mentat into data_layer
